File: backend-python/src/main/resources/org/shedlang/compiler/backends/python/modules/Stdlib/Platform/NonLocalReturn.py
class _NonLocalReturn(Exception):
    def __init__(self, value):
        super().__init__("non local return")
        self.value = value


def non_local_return(value):
    raise _NonLocalReturn(value)


def run(func, on_non_local_return):
    try:
        return func()
    except _NonLocalReturn as error:
        return on_non_local_return(error.value)

File: backend-javascript/src/main/resources/org/shedlang/compiler/backends/javascript/modules/Stdlib/Platform/NonLocalReturn.js
function nonLocalReturn(value) {
    throw {nonLocalReturn: true, value: value};
}

function run(func, onNonLocalReturn) {
    try {
        return func();
    } catch (error) {
        if (error.nonLocalReturn) {
            return onNonLocalReturn(error.value);
        } else {
            throw error;
        }
    }
}

exports.nonLocalReturn = nonLocalReturn;
exports.run = run;
